File: core/src/main/scala/com/electronwill/collections/SimpleBag.java
package com.electronwill.collections;

import java.util.Arrays;
import java.util.Iterator;

/**
 * A resizeable collection based on an array. The delete operation is in constant time because it
 * just moves the last element to fill the gap.
 *
 * @author TheElectronWill
 */
public class SimpleBag<E> extends AbstractBag<E> implements Bag<E> {
	private final int capacityIncrement;
	private Object[] array;
	private int size = 0;

	/**
	 * Constructs a new Bag with an initial capacity of ten and an increment of 2.
	 */
	public SimpleBag() {
		this(10, 2);
	}

	/**
	 * Creates a new SimpleBag with the given initialCapacity and a default increment of 2.
	 */
	public SimpleBag(int initialCapacity) {
		this(initialCapacity, 2);
	}

	/**
	 * Creates a new SimpleBag with the given initialCapacity and increment.
	 */
	public SimpleBag(int initialCapacity, int capacityIncrement) {
		array = new Object[initialCapacity];
		this.capacityIncrement = capacityIncrement;
	}

	public SimpleBag(E firstElement) {
		this(10, 2);
		array[size++] = firstElement;
	}

	public SimpleBag(E firstElement, int initialCapacity, int capacityIncrement) {
		this(initialCapacity, capacityIncrement);
		array[size++] = firstElement;
	}

	@Override
	public boolean add(E e) {
		if (size == array.length) {
			array = Arrays.copyOf(array, size + capacityIncrement);
		}
		array[size++] = e;
		return true;
	}

	@Override
	public void clear() {
		array = new Object[10];
	}

	@Override
	public boolean contains(Object o) {
		for (int i = 0; i < size; i++) {
			if (array[i].equals(o)) {
				return true;
			}
		}
		return false;
	}

	@Override
	public E get(int index) {
		if (index < size) {
			return (E)array[index];
		}
		throw new ArrayIndexOutOfBoundsException(index);
	}

	@Override
	public boolean isEmpty() {
		return size == 0;
	}

	@Override
	public Iterator<E> iterator() {
		return new BagIterator();
	}

	@Override
	public E remove(int index) {
		E element = (E)array[index];
		array[index] = array[--size];
		array[size] = null;
		return element;
	}

	@Override
	public boolean remove(Object o) {
		for (int i = 0; i < size; i++) {
			if (array[i].equals(o)) {
				remove(i);
				return true;
			}
		}
		return false;
	}

	@Override
	public int size() {
		return size;
	}

	@Override
	public Object[] toArray() {
		return Arrays.copyOf(array, size);
	}

	@Override
	public void compact() {
		array = Arrays.copyOf(array, size);
	}

	@Override
	public E tryGet(int index) {
		if (index < size) {
			return (E)array[index];
		}
		return null;
	}

	private class BagIterator implements Iterator<E> {
		private int pos = 0;
		private boolean removed;

		@Override
		public boolean hasNext() {
			return pos < size;
		}

		@Override
		public E next() {
			removed = false;
			return (E)array[pos++];
		}

		@Override
		public void remove() {
			if (removed) {
				throw new IllegalStateException("remove() already called for this element");
			}
			SimpleBag.this.remove(--pos);
			removed = true;
		}
	}
}

File: core/src/main/scala/com/electronwill/collections/SimpleBag.scala
package com.electronwill.collections

import scala.reflect.ClassTag

/**
 * @author TheElectronWill
 */
final class SimpleBag[A >: Null : ClassTag](initialCapacity: Int = 16) extends Bag[A] {
	private[this] var array = new Array[A](initialCapacity)
	private[this] var s: Int = 0

	override def size: Int = s
	override def apply(i: Int): A = array(i)
	override def remove(i: Int): Unit = {
		val last = array(s)
		array(i) = last
		array(s) = null
		s -= 1
	}

	override def -=(elem: A): SimpleBag.this.type = {
		var i = 0
		while (i < s && array(i) != elem) {
			i += 1
		}
		if (i != s) {
			remove(i)
		}
		this
	}
	override def +=(elem: A): this.type = {
		if (array.length == s) {
			array = grow(array, s + (s >> 1))
		}
		array(s) = elem
		s += 1
		this
	}
	override def iterator: MutableIterator[A] = new MutableIterator[A] {
		private[this] var i = 0
		private[this] val l = s

		override def hasNext: Boolean = i < l
		override def next(): A = {
			val v = array(i)
			i += 1
			v
		}
		override def remove(): Unit = {
			SimpleBag.this.remove(i)
		}
		override def insert(elem: A): Unit = {
			SimpleBag.this.+=(elem)
		}
	}
	override def compact(): Unit = {
		if (array.length > s) {
			array = shrink(array, s)
		}
	}
	override def clear(): Unit = {
		var i = 0
		while (i < s) {
			array(i) = null
			i += 1
		}
		s = 0
	}
}